src/services/refactors/extractSymbol.ts

```diff
@@ -265,24 +264,30 @@ namespace ts.refactor.extractSymbol {
     /**
      * getRangeToExtract takes a span inside a text file and returns either an expression or an array
      * of statements representing the minimum set of nodes needed to extract the entire span. This
-     * process may fail, in which case a set of errors is returned instead (these are currently
-     * not shown to the user, but can be used by us diagnostically)
+     * process may fail, in which case a set of errors is returned instead. These errors are shown to
+     * users if they have the provideRefactorNotApplicableReason option set.
      */
     // exported only for tests
-    export function getRangeToExtract(sourceFile: SourceFile, span: TextSpan, considerEmptySpans = true): RangeToExtract {
+    export function getRangeToExtract(sourceFile: SourceFile, span: TextSpan, invoked = true): RangeToExtract {
         const { length } = span;
-        if (length === 0 && !considerEmptySpans) {
+        if (length === 0 && !invoked) {
             return { errors: [createFileDiagnostic(sourceFile, span.start, length, Messages.cannotExtractEmpty)] };
         }
-        const cursorRequest = length === 0 && considerEmptySpans;
+        const cursorRequest = length === 0 && invoked;
+
+        const startToken = getTokenAtPosition(sourceFile, span.start);
+        const endToken = findTokenOnLeftOfPosition(sourceFile, textSpanEnd(span));
+        /* If the refactor command is invoked through a keyboard action it's safe to assume that the user is actively looking for
+        refactor actions at the span location. As they may not know the exact range that will trigger a refactor, we expand the
+        searched span to cover a real node range making it more likely that something useful will show up. */
+        const adjustedSpan = startToken && endToken && invoked ? getAdjustedSpanFromNodes(startToken, endToken, sourceFile) : span;

```

Comment:
I'm not sure I understand why tokens are involved.  Couldn't this just be `invoked && charCodeAt(span.end) === CharacterCodes.semicolon`?

Reply:
Oh, I guess you might need the tokens to rule out the case where you're in the header of a for loop?  But it doesn't look like `getAdjustedSpanFromNodes` does that?

Reply:
`getAdjustedSpan` *mostly* adjusts the span we search inside of to start at the beginning of `startToken` and end at the end of `endToken`.

We do this because `getParentNodeInSpan` finds the largest nodes contained within the selected span, containing `startToken` and `endToken`, which may be the same parent node.

`getAdjustedSpan` basically snaps the span to the tokens since the aforementioned search would fail if say the end of the span terminated halfway through `endToken`, so no parent of `endToken` would ever be in the span.

Reply:
Is there a test case covering the `for` case?  I would find that reassuring.

Reply:
I added a couple of for statement unit tests so it's apparent what works there, though this PR does not change the refactoring behavior there.

Reply:
Thanks!  I appreciate it.